### ISSUE TITLE
Unnecessary call to chebfun/size in fevalColumn

### ISSUE THREAD

**Issue by asgeirbirkis** [chebfun/chebfun#1600]
Looking at the profiler output when studying #1599, I noticed a lot of calls to `chebfun/size` inside `chebfun/feval/fevalColumn`. These are in [line 85](https://github.com/chebfun/chebfun/blob/fix-chebop-shortPulses/%40chebfun/feval.m#L85).

However, we're only really interested in the result of `chebfun/size` with the second argument equal to 2, which is just what `numColumns` returns (and `numColumns` gets called inside `chebfun/size`). By changing line 85 from
```
numCols = size(f, 2);
```
to
```
numCols = numColumns(f);
```

and running the profiler for `chebtest('chebfun')`, it turns out that it saved more than 13000 calls to `chebfun/size`, while all the tests still passed.

I'm wondering whether there's a reason for this call to `size` rather than `numColumns` before I issue a PR addressing it.

I'm also wondering about the line that follows in `fevalColumn`:
```
numFuns = numel(f.funs);
```
Will this ever be anything else than 1, as things with `numel(f) > 1` only live at the `chebfun` level?

**Comment by nickhale** [comment]
I agree that line 85 can be changed to `numCols = numColumns(f);`.

I don't understand your second question regarding `numFuns`. This method **is** at the `chebfun` level?

**Comment by nickhale** [comment]
Probably the 
```
wasTransposed = f.isTransposed;
if ( f.isTransposed )
    f = transpose(f);
end
```
block can be removed as well (I suspect this was written before we had the `numColumns()` method).

**Comment by asgeirbirkis** [comment]
@nickhale yes, we're working at the `chebfun` level, however, the line in question,
```
numFuns = numel(f.funs);
```
appears inside `columnFeval`. For array-valued chebfuns, isn't `numel(f.funs)` always equal to 1, and for actual quasimatrices, we're already looping through individual `funs` in lines 62-64 of the actual `feval` method:
```
for k = 1:numel(F)
    out{k} = columnFeval(F(k), x, varargin{:});
end
```

If I remove the `;` at the end of line 86, these are the outputs I get from running some experiments:
```
>> x = chebfun('x');
>> x(.5);
numFuns =
     1
>> xx = [x sin(x)];
>> xx(.5)
numFuns =
     1
ans =
    2.0000    0.9093
>> x = chebfun('x');
>> x(.5);
numFuns =
     1
>> xx = [x sin(x)];
>> xx(.5);
numFuns =
     1
>> xxt = xx.';
>> xxt(.5);
numFuns =
     1
>> xxq = quasimatrix(xx);
>> numel(xx)
ans =
     1
>> numel(xxq)
ans =
     2
>> xxq(.5)
numFuns =
     1
numFuns =
     1
```

**Comment by asgeirbirkis** [comment]
Bahh, I was being stupid here, this is dealing with individual pieces of a chebfun, not columns of a quasimatrix, two many numels going on, so ignore the last comment...

**Comment by asgeirbirkis** [comment]
@nickhale yes, it appears that the logic of dealing with transposed chebfuns can be simplified as well, so will include that in an upcoming pull request.

The final thing that I want to address are the repeated `isempty` checks; in `chebfun/feval` for example, we start by checking whether the chebfun is empty (fair enough), but then the isempty check happens again inside `numColumns`, even though we know that we're arriving from a method that's already checked it! So I propose we introduce an optional parameter to `numColumns` which allows it to by pass the `isempty` check, will give implementing it a go.

**Comment by nickhale** [comment]
I think the problem with isempty checking is far more wide-ranging than just this one instance. I'd suggest leaving that alone for now.

**Comment by asgeirbirkis** [comment]
Yeah, that was the conclusion I had come to after digging around the code for the last half an hour or so, these checks are happening all over the place (as we've known for some time), so better probably do deal with them properly rather than patching it for individual files.

**Comment by gradywright** [comment]
This is a bit tangent to the issue raised but has to deal with `columnFeval`.  There is a performance hit with the following code when `numFuns=1`.  

```
% Points within the domain:
for k = 1:numFuns
    I = ( xReal >= dom(k) ) & ( xReal < dom(k+1) );
    if ( any(I(:)) )
        % Evaluate the appropriate fun:
        out(I,:) = feval(funs{k}, x(I), varargin{:});
    end
end

% Points to the right of the domain:
I = ( xReal >= dom(end) );
if ( any(I(:)) )
    out(I,:) =  feval(funs{end}, x(I), varargin{:});
end
```
The hit comes from evaluating at many points over the domain and having to handle a point at the right end completely separately from the rest of the points.  This results in two calls to `feval` at the `tech` level when really only one is necessary.  Could we add some logic to detect when `numFuns=1` and simplify the code in this case?

**Comment by jrus** [comment]
@gradywright seems also worth making the extrapolated parts run at the same time as the left and right segments, e.g.:

```matlab
% If we have only one fun, evaluate directly
if (numFuns == 1)
    
    out = feval(funs{1}, x(I), varargin{:})

else
    % First interval and points to the left
    xReal = real(x);
    I = xReal < dom(2);
    if ( any(I(:)) )
        out(I,:) = feval(funs{1}, x(I), varargin{:});
    end

    % Points within the domain:
    for k = 2:numFuns-1
        I = ( xReal >= dom(k) ) & ( xReal < dom(k+1) );
        if ( any(I(:)) )
            % Evaluate the appropriate fun:
            out(I,:) = feval(funs{k}, x(I), varargin{:});
        end
    end

    % Points to the right of the domain:
    I = ( xReal >= dom(end-1) );
    if ( any(I(:)) )
        out(I,:) =  feval(funs{end}, x(I), varargin{:});
    end
end
```

**Comment by nickhale** [comment]
I just pushed to the `fix-simplify-feval-logic` incorporating the suggestions of @gradywright and @jrus. It didn't really have much of an effect on runtime of `chebtest`, but since it makes the single fun case more transparent, I suggest we keep it.

@asgeirbirkis - if you're happy with these modifications then I'll merge to PR.

**Comment by jrus** [comment]
@nickhale [your version](https://github.com/chebfun/chebfun/commit/0481f28ba6ade30eb0f3985056a0ee9975bfa82a) still is going to call feval separately for the last interval and the far right endpoint whenever there are at least 2 funs, and also is going to separately call feval on the first/last interval and any points extrapolated to the left/right (respectively). In general probably not a big performance hit, but if you had a tight loop which repeatedly evaluated a small number of points for various chebfuns of a few segments, the slight extra overhead could be noticeable.

**Comment by nickhale** [comment]
@jrus - If you want to be super picky; when x is real (which is the 90% case), then the evaluations at breakpoints (including end points) are not used at all, since these are extracted from `f.pointValues`.

However, dealing with all these different cases will probably (a) have very little performance gain, (b) complicate the code significantly.

**Comment by jrus** [comment]
Here’s a simpler one:
```matlab
% If we have only one fun, evaluate directly
if (numFuns == 1)

    out(:,:) = feval(funs{1}, x(:), varargin{:})

else
    % extend first and last interval to infinity
    dom = [-Inf, dom(2:end-1), Inf]
    
    % First interval and points to the left:
    xReal = real(x);

    for ( k = 1:numFuns )
        I = ( xReal >= dom(k) ) & ( xReal < dom(k+1) );
        if ( any(I(:)) )
            % Evaluate the appropriate fun:
            out(I,:) = feval(funs{k}, x(I), varargin{:});
        end
    end
end
```

**Comment by nickhale** [comment]
OK, that could work. I've pushed an implementation.

**Comment by jrus** [comment]
(Just went to walk the dog.)

Actually, thinking about it, if the domain gets extended like that, it’s probably not worth special casing the one fun case anymore. 

If it’s just:
```matlab
% extend first and last interval to infinity, to handle extrapolation
dom([1, end]) = [-Inf, Inf]

xReal = real(x);

for ( k = 1:numFuns )
    I = ( xReal >= dom(k) ) & ( xReal < dom(k+1) );
    if ( any(I(:)) )
        % Evaluate the appropriate fun:
        out(I,:) = feval(funs{k}, x(I), varargin{:});
    end
end
```
...then when there’s only one fun, there will still be a vacuous `(x >= -Inf) & (x < Inf)` check, but I doubt it’s ever a bottleneck.

**Comment by asgeirbirkis** [comment]
@nickhale , @jrus with your modifications, there is actually another simplification possible, at least in the case of the current state of the code (1349b2a).

The call to `numColumns` is only necessary to initalize the size of the output matrix, so can be moved to the inside of the else statement, line 117. Line 116 could then become
```
numCols = size(out, 2);
```
I'm running a chebtest on this now, will push it if it passes!

**Comment by nickhale** [comment]
@jrus - I'd prefer to leave the one fun case as separate (keep the simple things simple).

@asgeirbirkis - I also considered this, but doesn't `numCols` also appear on line 168?

**Comment by asgeirbirkis** [comment]
@nickhale yes, hence in the one fun case, we'll read it from the size of the output -- it should be the same as the number of columns from evaluating the fun.

Just finished running `chebtest('chebfun')` with this modification, it got rid of 15000 calls to `chebfun.numColumns` !

**Comment by jrus** [comment]
- I think the lrFlag thing is pretty unclear. It would be better to just have separate scalar values leftFlag and rightFlag, instead of checking the super obscure `if ( lrFlag(1) || lrFlag(3) ) % left` which takes reading way up the code to understand.

- This code is doing a really wonky extra loop. There’s no reason for the internal loop which doesn’t depend at all on k to be repeated k times:

    ```matlab
% Loop over the FUNs:
for k = 1:(numFuns + 1)
    index = ( x == dom(k) );
    if ( any(index) )
        % If a left or right flag has been passed, we reassign pointValues 
        % to be left/right values.
        if ( lrFlag(1) || lrFlag(3) ) % left
            for j = 1:numFuns
                f.pointValues(j+1,:) = get(funs{j}, 'rval');
            end
        elseif ( lrFlag(2) || lrFlag(4) ) % right
            for j = 1:numFuns
                f.pointValues(j,:) = get(funs{j}, 'lval');
            end
        end
        pointValues = repmat(f.pointValues(k,:), sum(index), 1);
        out(index,:) = pointValues;
    end
end
```

    Actually it’s really wonky in general. Why is part of f.pointValues being mutated? Shouldn’t that be an internal feature of `f`? That is, if I look at `f.pointValues` and then I evaluate f to the right of each interval, should `f.pointValues` really change?

**Comment by jrus** [comment]
Is there any way to vectorize a call to `get(funs{j}, 'rval')`, so that it will return an array of rvals for a bunch of different values of `j`? I’m not all that familiar with matlab's cell/array distinction or all the relevant functions in the standard library.

**Comment by jrus** [comment]
Okay, I’m pretty sure it’s possible to do something like this:

```matlab
[x_in_dom, dom_indices] = ismember(x, dom)
if ( any(x_in_dom) )
    pointValues = f.pointValues
    if ( leftFlag )
        pointValues(2:end,:) = cellfun(@(fun) get(fun, 'rval'), funs);
    elseif ( rightFlag )
        pointValues(1:end-1,:) = cellfun(@(fun) get(fun, 'rval'), funs);
    end
    out(x_in_dom,:) = pointValues(dom_indices,:)
end
```

**Comment by gradywright** [comment]
Thanks @asgeirbirkis, @nickhale, and @jrus for looking into this issue, and for providing a fix in the `fix-simplify-feval-logic` branch.  With the current code in this branch we get quite a speed-up on certain operations.  For example, consider the following on the development branch:

```
>> f = chebfun2(@(x,y) cos(pi*exp(2*x+4*tanh(y))));
>> tic, max2(f), toc

ans =

    1.0000

Elapsed time is 4.005418 seconds.
```
On the `fix-simplify-feval-logic` branch we now get
```
>> f = chebfun2(@(x,y) cos(pi*exp(2*x+4*tanh(y))));
>> tic, max2(f), toc

ans =

    1.0000

Elapsed time is 1.816906 seconds.
```
More than two times a speed up!

**Comment by nickhale** [comment]
@jrus - How about this?
```
[x_in_dom, dom_indices] = ismember(x, dom);
if ( any(x_in_dom) )
    pointValues = f.pointValues;
    if ( leftFlag )
        pointValues(2:end,:) = get(f, 'rval-local');
    elseif ( rightFlag )
        pointValues(1:end-1,:) = get(f, 'lval-local');
    end
    dom_indices(~x_in_dom) = [];
    out(x_in_dom,:) = pointValues(dom_indices,:);
end
```

**Comment by jrus** [comment]
Ah right, the lingering zeroes in dom_indices in my version won’t do. That one looks good.

**Comment by asgeirbirkis** [comment]
@gradywright Nice, that's a pretty nice side effect from originally trying to speed up solving IVPs!

**Comment by jrus** [comment]
One last question: are there ever cases where a chebfun of many segments is evaluated only (or predominantly) at the endpoints of individual funs? If so, those might benefit from rearranging the code to look up the pointvalues first, and then skip those elements of `x` when doing the general feval on each fun. On the other hand, if such uses are rare, it might not be worth bothering.

**Comment by jrus** [comment]
**Oh!**. I was confused about the meaning of `left` / `right`. Those mean “take the value using the interval to the left/right of the point”, not “use the value at the left/right endpoints of each interval”. That should be maybe clarified in some documentation somewhere, if it isn’t already.

Anyhow, @nickhale, how’s this:

```matlab
%% POINTVALUES:
% If the evaluation point corresponds to a breakpoint, we get the value from
% pointValues. 

[x_in_dom, dom_indices] = ismember(x, dom);
if ( any(x_in_dom) )
    
    % Fetch the point values. When left / right flag is set, fetch the
    % appropriate endpoint values from the individual funs.
    if ( leftFlag )
        pointValues = [f.pointValues(1), get(f, 'rval-local')]
    elseif ( rightFlag )
        pointValues = [get(f, 'lval-local'), f.pointValues(end)]
    else
        pointValues = f.pointValues
    end
    
    % overwrite the output values for any values of x at the breakpoints
    out(x_in_dom,:) = pointValues(dom_indices(x_in_dom),:);
end
```

Along with earlier up the file:
```matlab
%% LEFT AND RIGHT LIMITS:
% Deal with feval(f, x, 'left') and feval(f, x, 'right'):
leftFlag = 0;
rightFlag = 0;
if ( nargin > 2 )
    lr = varargin{1};
    leftFlag = any(strcmpi(lr, {'left', '-'}));
    rightFlag = any(strcmpi(lr, {'right', '+'}));
    if ( ~(leftFlag | rightFlag))
        if ( ischar(lr) )
            error('CHEBFUN:CHEBFUN:feval:leftRightChar',...
                'Unknown input argument "%s".', lr);
        else
            error('CHEBFUN:CHEBFUN:feval:leftRight', 'Unknown input argument.');
        end
    end
end
```

**Comment by jrus** [comment]
Here’s a gist of the full file and a diff showing how I’d do it:
https://gist.github.com/jrus/7760a42347aa70c7aa05

I switched to camelCase variable names.

**Comment by jrus** [comment]
Okay, another question. (There are a lot of edge cases here, it turns out.)

If someone calls e.g. `feval(cfun, [0 + 1i], 'left')` on a chebfun with a breakpoint at 0, should that evaluate on the interval to the left of 0? Currently it still evaluates that from the right.

Likewise, should `feval(cfun, [0 + 1i])` return the average value approaching from the left / right?

**Comment by jrus** [comment]
I created a related issue at https://github.com/chebfun/chebfun/issues/1606 including both a report of what seems to be a bug, and then in two follow-up comments a suggestion for speeding up chebfun.feval, with a code sketch.
